nifi-nar-bundles/nifi-asana-bundle/nifi-asana-processors/src/main/java/org/apache/nifi/processors/asana/GetAsanaObject.java

```diff
@@ -0,0 +1,457 @@
+        flowFiles.put(AsanaObjectState.NEW, newItems);
+        flowFiles.put(AsanaObjectState.UPDATED, updatedItems);
+        flowFiles.put(AsanaObjectState.REMOVED, removedItems);
+
+        List<AsanaObject> allObjects = new ArrayList<>();
+
+        AsanaObject nextObject;
+        while ((nextObject = objectFetcher.fetchNext()) != null) {
+            allObjects.add(nextObject);
+        }
+
+        Map<AsanaObjectState, List<AsanaObject>> allObjectsByState = allObjects.stream()
+                .collect(groupingBy(AsanaObject::getState));
+
+        if (batchSize == 1) {
+            allObjectsByState
+                    .forEach((asanaObjectState, asanaObjects) -> asanaObjects.forEach(
+                            asanaObject -> {
+                                final Map<String, String> attributes = new HashMap<>(2);
+                                attributes.put(CoreAttributes.MIME_TYPE.key(), ContentType.APPLICATION_JSON.getMimeType());
+                                attributes.put(ASANA_GID, asanaObject.getGid());
+                                FlowFile flowFile = createFlowFileWithStringPayload(session, asanaObject.getContent());
+                                flowFile = session.putAllAttributes(flowFile, attributes);
+                                flowFiles.get(asanaObject.getState()).add(flowFile);
+                            }
+                    ));
+        } else {
+            allObjectsByState
+                    .forEach((asanaObjectState, asanaObjects) -> partition(asanaObjects, batchSize).forEach(
+                            asanaObjectsInPartition -> {
+                                FlowFile flowFile = createFlowFileWithStringPayload(session, format("[%s]",
+                                        asanaObjectsInPartition.stream().map(AsanaObject::getContent)
+                                                .collect(joining(","))));
+                                flowFile = session.putAllAttributes(flowFile,
+                                        singletonMap(CoreAttributes.MIME_TYPE.key(),
+                                                ContentType.APPLICATION_JSON.getMimeType()));
+                                flowFiles.get(asanaObjectState).add(flowFile);
+                            }
+                    ));
+        }
+
+        if (flowFiles.values().stream().allMatch(Collection::isEmpty)) {
+            context.yield();
+            getLogger().debug("Yielding, as there are no new FlowFiles.");
+        } else {
+            session.transfer(newItems, REL_NEW);
+            session.transfer(updatedItems, REL_UPDATED);
+            session.transfer(removedItems, REL_REMOVED);
+            session.commitAsync();
+        }
+        Map<String, String> state = objectFetcher.saveState();
+        try {
+            persistState(state, context);
+        } catch (IOException e) {
+            throw new ProcessException(e);
+        }
```

Comment:
Tried to apply this suggestion in df3c96e . but since `[AsanaProjectEventFetcher](https://github.com/apache/nifi/blob/126b1b5aaf2158e6c1f68854bb1058f0971d3578/nifi-nar-bundles/nifi-asana-bundle/nifi-asana-processors/src/main/java/org/apache/nifi/processors/asana/utils/AsanaProjectEventFetcher.java)` generates new state even if there are no new `FlowFiles` are generated, it would break, because the new state would not be persisted.

Added 126b1b5 to address this issue. Only parts of your suggestion have been applied.